### DISCUSSION TITLE
(non)blocking both way communication via serial

### DISCUSSION THREAD

**Discussion by emcek** [micropython/micropython#10474] [Using MicroPython]
I looking for (at best) non-blocking both way communication Pico <=> PC via serial. I need send something to Pico (i.e. print at LCD) and allow Pico to send to PC events from LCD touch screen or onboard buttons.

Found:
[belay](https://github.com/BrianPugh/belay) looks like communication is initiate by PC only: printing something at LCD or return value from some function executed at Pico.

I guess it should be already written by somebody, I don't need reinvented wheel....

**Comment by peterhinch** [comment]
See [this demo](https://github.com/peterhinch/micropython-async/blob/master/v3/as_demos/auart.py).

**Comment by emcek** [comment]
I try set `UART` to `0` and `1`, since I don't have 4 at my **Pico**. But nothing happend after running `test()`. Should I split code and put at PC (and replace `import uasyncio as asyncio` to `import asyncio as asyncio`)?

**Reply by peterhinch** [reply]
The demo should definitely work on a Pico, but you'll have to specify pins when you instantiate the UARTs as the Pico offers a choice. For example UART0 TX can be on GP 0, 12 or 16. Note that the demo uses a loopback, based on a single UART with TX and RX linked. This is to illustrate concurrency. It can readily be adapted for two UARTs with the transmitting UART's TX linked to the receiving UARTs RX.

When you come to do the PC side of the link you may need to find an alternative library. I haven't tested the `uasyncio` stream interface on the Unix build.

**Reply by cve2022** [reply]
@peterhinch , a (possibly stupid) question: if one does not specify the pins, shouldn't it just route (or continue routing) the UART0 through the USB port of the Pico?

**Reply by robert-hh** [reply]
> if one does not specify the pins, shouldn't it just route (or continue routing) the UART0 through the USB port of the Pico?

No. In that case default pins are used. GP0 and 1 for UART0, GP 4 and 5 for UART1.

**Comment by cve2022** [comment]
Hello @emcek

I am facing a similar situation as yours, I still did not had the time to work on the code. I plan using the example linked by @peterhinch as a starting point, but I am going to use sys.stdin and sys.stdout instead of directly tampering with the UART to make a (hopefully) smooth transition between REPL and application and vice-versa during the development of my code.

Are you using Thonny? If affirmative, then it may be the cause of your problem (from what I have read around here, Thonny do some weird things that may interfere with the normal operation of the MP application). I suggest you try [rshell ](https://pypi.org/project/rshell/) or [mpremote](https://pypi.org/project/mpremote/).

HTH.
